File: services/template-service/templates_storage/maestro-curated/ml-pipeline-orchestrator/src/infrastructure/resilience/retry.py
"""Retry with exponential backoff pattern.

Automatically retries failed operations with increasing delays.
Per ADR-006 resilience requirements.
"""

import asyncio
import logging
from typing import Callable, TypeVar, Type

logger = logging.getLogger(__name__)

T = TypeVar('T')


class RetryExhaustedError(Exception):
    """Raised when all retry attempts are exhausted"""
    pass


async def retry_with_backoff(
    func: Callable[..., T],
    max_retries: int = 3,
    initial_delay: float = 1.0,
    backoff_factor: float = 2.0,
    max_delay: float = 60.0,
    retryable_exceptions: tuple[Type[Exception], ...] = (Exception,),
    operation_name: str = "operation",
) -> T:
    """Retry an async function with exponential backoff.

    Args:
        func: Async function to retry
        max_retries: Maximum number of retry attempts
        initial_delay: Initial delay in seconds
        backoff_factor: Multiplier for delay after each retry
        max_delay: Maximum delay between retries
        retryable_exceptions: Tuple of exception types to retry
        operation_name: Name for logging

    Returns:
        Result from successful function call

    Raises:
        RetryExhaustedError: If all retries exhausted
        Any non-retryable exception
    """
    delay = initial_delay
    last_exception = None

    for attempt in range(max_retries + 1):
        try:
            result = await func()
            if attempt > 0:
                logger.info(
                    f"Operation '{operation_name}' succeeded on attempt {attempt + 1}"
                )
            return result

        except retryable_exceptions as e:
            last_exception = e

            if attempt >= max_retries:
                logger.error(
                    f"Operation '{operation_name}' failed after {max_retries + 1} attempts: {e}"
                )
                raise RetryExhaustedError(
                    f"Operation '{operation_name}' failed after {max_retries + 1} attempts"
                ) from e

            wait_time = min(delay, max_delay)
            logger.warning(
                f"Operation '{operation_name}' failed (attempt {attempt + 1}/{max_retries + 1}), "
                f"retrying in {wait_time:.1f}s: {e}"
            )

            await asyncio.sleep(wait_time)
            delay *= backoff_factor

        except Exception as e:
            # Non-retryable exception
            logger.error(f"Operation '{operation_name}' failed with non-retryable error: {e}")
            raise

    # Should never reach here
    raise RetryExhaustedError(
        f"Operation '{operation_name}' failed after {max_retries + 1} attempts"
    ) from last_exception

File: services/template-service/templates_storage/maestro-curated/sunday-com/backend/src/infrastructure/resilience/retry.ts
/**
 * Retry with Exponential Backoff Pattern
 *
 * Automatically retries failed operations with increasing delays.
 * Prevents overwhelming failing services while giving them time to recover.
 *
 * ADR-006 Compliant: Resilience Patterns
 */

export interface RetryConfig {
  maxRetries: number;           // Maximum number of retry attempts
  initialDelay: number;         // Initial delay in milliseconds
  maxDelay: number;             // Maximum delay in milliseconds
  backoffMultiplier: number;    // Multiplier for exponential backoff (default: 2)
  retryableErrors?: string[];   // List of retryable error types/messages
}

export interface RetryMetrics {
  attempts: number;
  totalDelay: number;
  success: boolean;
  finalError?: Error;
}

/**
 * Sleep for specified milliseconds
 */
function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * Check if error is retryable
 */
function isRetryableError(error: Error, retryableErrors?: string[]): boolean {
  if (!retryableErrors || retryableErrors.length === 0) {
    return true; // Retry all errors by default
  }

  return retryableErrors.some(
    errorType =>
      error.name === errorType ||
      error.message.includes(errorType)
  );
}

/**
 * Execute function with retry logic and exponential backoff
 */
export async function withRetry<T>(
  fn: () => Promise<T>,
  config: RetryConfig,
  context?: string
): Promise<T> {
  const {
    maxRetries,
    initialDelay,
    maxDelay,
    backoffMultiplier = 2,
    retryableErrors
  } = config;

  let lastError: Error;
  let delay = initialDelay;
  let totalDelay = 0;

  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      // First attempt or retry after delay
      if (attempt > 0) {
        await sleep(delay);
        totalDelay += delay;

        console.log(
          `Retry attempt ${attempt}/${maxRetries}${context ? ` for ${context}` : ''} after ${delay}ms delay`
        );
      }

      const result = await fn();

      if (attempt > 0) {
        console.log(
          `Retry successful after ${attempt} attempts (total delay: ${totalDelay}ms)${context ? ` for ${context}` : ''}`
        );
      }

      return result;
    } catch (error) {
      lastError = error as Error;

      // Check if error is retryable
      if (!isRetryableError(lastError, retryableErrors)) {
        throw new Error(
          `Non-retryable error${context ? ` for ${context}` : ''}: ${lastError.message}`
        );
      }

      // Last attempt failed
      if (attempt === maxRetries) {
        throw new Error(
          `Max retries (${maxRetries}) exceeded${context ? ` for ${context}` : ''}. Last error: ${lastError.message}`
        );
      }

      // Calculate next delay with exponential backoff
      delay = Math.min(delay * backoffMultiplier, maxDelay);
    }
  }

  throw lastError!;
}

/**
 * Retry decorator for class methods
 */
export function Retry(config: RetryConfig) {
  return function (
    target: any,
    propertyKey: string,
    descriptor: PropertyDescriptor
  ) {
    const originalMethod = descriptor.value;

    descriptor.value = async function (...args: any[]) {
      return withRetry(
        () => originalMethod.apply(this, args),
        config,
        `${target.constructor.name}.${propertyKey}`
      );
    };

    return descriptor;
  };
}

/**
 * Default retry configurations for common scenarios
 */
export const defaultRetryConfigs: Record<string, RetryConfig> = {
  database: {
    maxRetries: 3,
    initialDelay: 1000,    // 1 second
    maxDelay: 10000,       // 10 seconds
    backoffMultiplier: 2,
    retryableErrors: [
      'ECONNREFUSED',
      'ETIMEDOUT',
      'ECONNRESET',
      'EPIPE'
    ]
  },
  cache: {
    maxRetries: 2,
    initialDelay: 500,     // 0.5 seconds
    maxDelay: 5000,        // 5 seconds
    backoffMultiplier: 2,
    retryableErrors: [
      'ECONNREFUSED',
      'ETIMEDOUT'
    ]
  },
  external_api: {
    maxRetries: 3,
    initialDelay: 2000,    // 2 seconds
    maxDelay: 30000,       // 30 seconds
    backoffMultiplier: 2,
    retryableErrors: [
      'ECONNREFUSED',
      'ETIMEDOUT',
      'ENOTFOUND',
      '429',  // Too Many Requests
      '503',  // Service Unavailable
      '504'   // Gateway Timeout
    ]
  },
  file_operation: {
    maxRetries: 2,
    initialDelay: 1000,
    maxDelay: 5000,
    backoffMultiplier: 2,
    retryableErrors: [
      'EBUSY',
      'EMFILE',
      'ENFILE'
    ]
  }
};

/**
 * Retry with jitter to prevent thundering herd
 */
export async function withRetryAndJitter<T>(
  fn: () => Promise<T>,
  config: RetryConfig,
  jitterFactor: number = 0.1,
  context?: string
): Promise<T> {
  const configWithJitter = {
    ...config,
    initialDelay: config.initialDelay * (1 + Math.random() * jitterFactor)
  };

  return withRetry(fn, configWithJitter, context);
}
